[GH-31650] Race Condition When Sending Multiple Replies to a Post

berylxo opened the pull request:
#### Summary
This PR fixes a race condition in the thread update logic when multiple replies are posted simultaneously.
Previously, posting replies at the same time could cause a duplicate key violation in the Threads table.
The fix ensures that concurrent inserts/updates for threads are handled safely via ON CONFLICT logic.

##### QA Steps

1. `Create a New Post`

```bash
curl --request POST \
  --url http://localhost:8065/api/v4/posts \
  --header 'Accept: application/json' \
  --header "Authorization: Bearer <token>" \
  --header 'Content-Type: application/json' \
  --data '{
    "channel_id": "<channel_id>",
    "message": "This is a test message."
  }'

```
2. `Copy message_id from the response`

3. `Simulate simultaneous replies`
```bash
curl --request POST \
  --url http://localhost:8065/api/v4/posts \
  --header 'Accept: application/json' \
  --header "Authorization: Bearer <token>" \
  --header 'Content-Type: application/json' \
  --data "{
    \"channel_id\": \"<channel_id>\",
    \"message\": \"This is a test reply 1.\",
    \"root_id\": \"<message_id>\"
  }" &

curl --request POST \
  --url http://localhost:8065/api/v4/posts \
  --header 'Accept: application/json' \
  --header "Authorization: Bearer <token>" \
  --header 'Content-Type: application/json' \
  --data "{
    \"channel_id\": \"<channel_id>\",
    \"message\": \"This is a test reply 2.\",
    \"root_id\": \"<message_id>\"
  }" &
```
4. `Before fix behavior: one reply is posted and server returns duplicate key error for the other post`
```bash
{"id":"app.post.save.app_error","message":"Unable to save the Post.","detailed_error":"update thread from posts failed: pq: duplicate key value violates unique constraint \"threads_pkey\"","request_id":"5pkz4gb6xid6fdfj688znycrae","status_code":500}
{"id":"qxwa9pse5b8z5ghjqtbtwszg7y","create_at":1756729989763,"update_at":1756729989763,"edit_at":0,"delete_at":0,"is_pinned":false,"user_id":"oqpbboxtzff8fdurz5efm9cqpa","channel_id":"dqc3xhyyjtrgiewyqkzuur84qo","root_id":"a4bc1dmxwp8z5kztgm67y8ahny","original_id":"","message":"This is a test reply 2.","type":"","props":{},"hashtags":"","file_ids":[],"pending_post_id":"","remote_id":"","reply_count":1,"last_reply_at":0,"participants":null,"metadata":{}}
```
5. ` After Fix: Both replies are posted`
```bash
{"id":"k9fdbcet7trpbfcqugqw1zedzy","create_at":1756731725537,"update_at":1756731725537,"edit_at":0,"delete_at":0,"is_pinned":false,"user_id":"oqpbboxtzff8fdurz5efm9cqpa","channel_id":"dqc3xhyyjtrgiewyqkzuur84qo","root_id":"f7srtyfr5b8jte9j6i3yexo3ke","original_id":"","message":"This is a test reply 1.","type":"","props":{},"hashtags":"","file_ids":[],"pending_post_id":"","remote_id":"","reply_count":2,"last_reply_at":0,"participants":null,"metadata":{}}
{"id":"sgxwfd1ii7ffjbtu4twdjgs9ca","create_at":1756731725537,"update_at":1756731725537,"edit_at":0,"delete_at":0,"is_pinned":false,"user_id":"oqpbboxtzff8fdurz5efm9cqpa","channel_id":"dqc3xhyyjtrgiewyqkzuur84qo","root_id":"f7srtyfr5b8jte9j6i3yexo3ke","original_id":"","message":"This is a test reply 2.","type":"","props":{},"hashtags":"","file_ids":[],"pending_post_id":"","remote_id":"","reply_count":2,"last_reply_at":0,"participants":null,"metadata":{}}
```

#### Ticket Link
Fixes [mattermost/mattermost#31650](https://github.com/mattermost/mattermost/issues/31650)

#### Release Note

```release-note
Fixed a race condition when updating thread metadata that could cause "duplicate key value violates unique constraint threads_pkey" errors under concurrent replies.
```

berylxo committed:
fix: update query to allow multiple replies simultaneosly

berylxo committed:
fix: participants aggregation logic
- Use COALESCE with '[]'::jsonb to guarantee non-null result.
  - Merge existing and new participants via jsonb_array_elements + UNION.
  - Apply DISTINCT in jsonb_agg to avoid duplicate entries.
- Prevents cases where participants could be stored as
  instead of an empty JSON array, ensuring consistency with
  model.StringArray serialization in Go.

berylxo committed:
threads: remove forced empty participants initialization

Previously, participants was being set to an empty array when nil. This is
unnecessary since the SQL upsert already uses COALESCE to handle NULL values.
Removing this ensures that the participants field reflects the actual state
and avoids masking cases where no participants exist.

hanzei reviewed (changes requested):
Thanks for the PR 👍 
The change looks fine on the code level.

Would you please also add a test that covers the changes? Please ensure that the test doesn't pass with `master`.

berylxo committed:
test: Added a new test (PostStoreConcurrentReplies) to verify that multiple replies saved in parallel succeed and that the thread's ReplyCount and Participants are updated properly.

berylxo committed:
concurrent replies test

berylxo committed:
test: add test on concurrent replies to the same thread

berylxo committed:
cleanup comments

BerylCAtieno commented:
@hanzei @agarciamontoro 

I've pushed comprehensive tests to validate the concurrent thread update fix in `testConcurrentReplies()`. The tests are designed to fail on master and pass with the changes.

## Test Coverage:

- **Test 1**: Verifies no duplicate key violations occur during concurrent replies to the same thread
- **Test 2**: Validates thread metadata (`replyCount`, `participants`, `LastReplyAt`) updates correctly under concurrency
- **Test 3**: Ensures all concurrent replies are properly persisted and retrievable from the database

### To validate the fix:

- Run `testConcurrentReplies` on master branch - should see duplicate key constraint violations
- Run the same tests with this PR - should pass cleanly

agarciamontoro reviewed (changes requested):
Thank you for your work, @BerylCAtieno 🎉 

There's a couple of things I believe we need to change, but other than that this is a great solution! I'm also not very concerned about performance, since the `ON CONFLICT` path will not happen frequently.

A more generic ask: can we repeat the test multiple times (let's say, 10), so that we can be more confident the bug is gone? Having multiple concurrent replies is already a good approach, but I would like to be extra sure here.

BerylCAtieno commented on server/channels/store/sqlstore/post_store.go:
```diff
@@ -3135,17 +3135,26 @@ func (s *SqlPostStore) updateThreadsFromPosts(transaction *sqlxTxWrapper, posts
 			}
 			// no metadata entry, create one
-			if _, err := transaction.NamedExec(`INSERT INTO Threads
+			if _, err := transaction.Exec(`INSERT INTO Threads
```
My only reason for using `Exec` is that I found it to be slightly less verbose given the complexity of the query. I will edit to use `NamedExec` and push alongside the other suggested changes.

BerylCAtieno replied:
I have tried to implement `NamedExec` and remembered why I changed it in the first place. I have tried several approaches but I keep getting `syntax error at ':'` . This occurs because the sqlx `NamedExec` implementation is struggling with parameter binding in the context of PostgreSQL's json functions and EXCLUDED table references. When the named parameters are being translated to positional parameters internally, the translation is failing to properly handle the complex nesting of the json operations and the special EXCLUDED table reference.

Apparently, this is a common limitation when mixing `sqlx` named parameters with certain PostgreSQL-specific features, especially in complex UPSERT queries that use json operations. The parameter binding mechanism gets confused about whether ':' is part of a named parameter or part of the PostgreSQL json operation syntax.

BerylCAtieno commented on server/channels/store/sqlstore/post_store.go:
```diff
@@ -3135,17 +3135,26 @@ func (s *SqlPostStore) updateThreadsFromPosts(transaction *sqlxTxWrapper, posts
+							),
+							'[]'::jsonb
+						),
+					ThreadTeamId = EXCLUDED.ThreadTeamId,
+					ChannelId = EXCLUDED.ChannelId
```
You're right. These values are part of the thread's core identity and shouldn't change. I will remove the two line in my PR update

berylxo committed:
fix: increment reply count for each reply

berylxo committed:
test:
- I've removed the redundant assertion about LastReplyAt being greater than the root post's creation time
- I've added test iterations to run the test several times, keeping same concurrent reply logic

hanzei reviewed (changes requested):
There are a lot of changes in your branch that don't belong there. Please update the `master` branch of your fork and then run `git rebase master` from your feature branch to remove them.

Happy to help, if needed!

hanzei commented on server/channels/store/storetest/post_store.go:
There are some linter errors:

```
/__w/mattermost/mattermost/server/channels/store/storetest/post_store.go:5737:6: for loop can be modernized using range over int
/__w/mattermost/mattermost/server/channels/store/storetest/post_store.go:5782:7: for loop can be modernized using range over int
```

hanzei commented:
And the tests are failing. Can you please take a look? https://github.com/mattermost/mattermost/actions/runs/17979974633/job/51151154473?pr=33825

berylxo committed:
range syntax for the loop

berylxo committed:
fix: prevent lost updates when incrementing thread reply count

Previously, the UPSERT for Threads used
in the ON CONFLICT clause. Under concurrent inserts, multiple transactions
could read the same old value and overwrite each other’s updates, resulting
in fewer replies being counted (e.g. 8 replies expected but only 4 stored).

This change replaces the fixed  increment with
, ensuring that
each conflicting insert contributes its own increment. This avoids lost
updates and makes concurrent replies update metadata correctly.

berylxo committed:
increased concurrent replies and iterations for stress testing

BerylCAtieno commented:
@hanzei @agarciamontoro I have discovered that the approach I used here does not effectively fix the race condition. It fixes the duplicate key error, but race condition still exists when updating thread metadata. I am exploring alternative approaches to fix the problem and will update you when ready for review.

berylxo committed:
Fix: Thread metadata race condition on concurrent replies
The fix employs a combination of database-level locking and collision avoidance to guarantee transactional integrity for both thread creation and metadata updates

berylxo committed:
test: increase number of concurrent replies and iterations

berylxo committed:
fix race condition in thread metadata update

berylxo committed:
fix race condition with db safeguards and in app update logic

berylxo committed:
fix: lint error

berylxo committed:
style: delete trailing line

agarciamontoro commented:
I've been reading on row-level locks, since I'm not familiar with them, but I still don't have an opinion fully formed on this. Will continue the review on Monday, since I'm out tomorrow. Thank you for your patience, @BerylCAtieno!

hanzei reviewed (dismissed):
Thanks for the continuous updates on the PR @BerylCAtieno 🎉 

Would you be open to merging `master` into your branch and fixing the merge conflict? Thanks!

berylxo committed:
Merge branch 'master' into MM-31650_race-condition-multiple-replies

agarciamontoro reviewed (changes requested):
This PR fell under my radar, sorry, @BerylCAtieno! I see that you need to merge `master` again, but your tests seem to be passing now. Could you merge `master` and ask for my review again when that's done? Thank you!

berylxo committed:
Merge branch 'master' into MM-31650_race-condition-multiple-replies

berylxo committed:
Merge branch 'master' into MM-31650_race-condition-multiple-replies

berylxo committed:
refactor: replace raw insert with query builder

berylxo committed:
fix shadowing lint error

berylxo committed:
Merge branch 'master' into MM-31650_race-condition-multiple-replies

hanzei commented:
If the flaky test is not related to your PR, that you are good to ignore it :stuck_out_tongue:

carlisgg reviewed (approved):
pending all test pass, it LGTM

berylxo committed:
Merge branch 'master' into MM-31650_race-condition-multiple-replies
